src/python/bcc/remote/libremote.py

```diff
@@ -0,0 +1,225 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import base64
+import re
+import ctypes as ct
+from shell import ShellRemote
+from adb import AdbRemote
+
+def get_remote_cls(cls_name):
+    cls_name = cls_name.capitalize() + 'Remote'
+    cls = globals()[cls_name]
+    return cls
+
+class LibRemote(object):
+    def __init__(self, remote_name, remote_arg=None):
+        # Get the <class>Remote class object
+        cls = get_remote_cls(remote_name)
+
+        # Cache of maps, format: map_cache[map_fd] = { 'key1': 'value1', .. }
+        self.nkey_cache = {}
+        self.map_cache = {}
+
+        # Create the remote connection
+        self.remote = cls(remote_arg)
+
+    def _remote_send_command(self, cmd):
+        """
+        Return: a tuple containing return code and output list.
+        """
+        ret = self.remote.send_command(cmd)
+        if not ret:
+            return (-1, [])
+
+        if ret[0].startswith('Command not recognized'):
+            print('Command not recognized! cmd: {}'.format(cmd))
+            return (-1, [])
+
+        # Assume success if first list element doesn't have ret=
+        if not 'ret=' in ret[0]:
+            return (0, ret)
+
+        m = re.search("ret=(\-?\d+)", ret[0])
+        if m == None:
+            print('Bad return string for cmd {}'.format(cmd))
+            return (-1, [])
+
+        return (int(m.group(1)), ret)
+
+    def available_filter_functions(self, tracefs):
+        cmd = "GET_AVAIL_FILTER_FUNCS {}".format(tracefs)
+        ret = self._remote_send_command(cmd)
+        return ret[0] if ret[0] < 0 else ret[1]
+
+    def kprobes_blacklist(self, tracefs):
+        cmd = "GET_KPROBES_BLACKLIST {}".format(tracefs)
+        ret = self._remote_send_command(cmd)
+        return ret[0] if ret[0] < 0 else ret[1]
+
+    def get_trace_events(self, tracefs, cat):
+        cmd = "GET_TRACE_EVENTS {} {}".format(tracefs, cat)
+        ret = self._remote_send_command(cmd)
+        return ret[0] if ret[0] < 0 else ret[1]
+
+    def get_trace_events_categories(self, tracefs):
+        cmd = "GET_TRACE_EVENTS_CATEGORIES {}".format(tracefs)
+        ret = self._remote_send_command(cmd)
+        return ret[0] if ret[0] < 0 else ret[1]
+
+    def bpf_attach_tracepoint(self, fd, cat, tpname):
+        cmd = "BPF_ATTACH_TRACEPOINT {} {} {}".format(fd, cat, tpname)
+        ret = self._remote_send_command(cmd)
+        return ret[0]
+
+    def bpf_attach_kprobe(self, fd, t, evname, fnname):
+        cmd = "BPF_ATTACH_KPROBE {} {} {} {}".format(fd, t, evname, fnname)
+        ret = self._remote_send_command(cmd)
+        return ret[0]
+
+    def bpf_detach_kprobe(self, evname):
+        # this is hard to do since a signal needs to be delivered if
```

Comment:
This issue is the same as the perf reader polling issue or it could be the same thing. Either way, why isn't this fixed/cleaned up?